fixed: asyncpg connection params are a namedtuple

privatwolke opened the pull request:
Follow-up on the apparently abbandonned https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2114.
Closes #2114

# Description

The asyncpg instrumentation attempts to fall back on using the database name as the span name in case the first argument to the instrumented method is falsey.

Any attempt to do so results in errors simliar to this (truncated traceback):
```
Traceback (most recent call last):
  File "/opt/venv/lib/python3.12/site-packages/asyncpg/pool.py", line 597, in fetch
    return await con.fetch(
           ^^^^^^^^^^^^^^^^
  File "/opt/venv/lib/python3.12/site-packages/opentelemetry/instrumentation/asyncpg/__init__.py", line 154, in _do_execute
    name = args[0] if args[0] else params.get("database", "postgresql")
                                   ^^^^^^^^^^
AttributeError: 'ConnectionParameters' object has no attribute 'get'
```

This has probably never worked since asyncpg defines the `_params` attribute as an instance of `ConnectionParams`
(https://github.com/MagicStack/asyncpg/blob/master/asyncpg/connection.py#L62) which is a NamedTuple instance and thus don't define `get`. The proper way of safely accessing properties on a NamedTuple is using `getattr`.

The only case that I've actually found which triggers this branch is if the supplied query is an empty string. This is something that causes an `AttributeError` for `Connection.execute` but is fine for `fetch()`, `fetchval()`, `fetchrow()` and `executemany()`.

The tests have been expanded to check these cases. Also, more status code validation has been added where it was missing.

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

- [x] docker-tests for asyncpg instrumentation extended and passed

# Does This PR Require a Core Repo Change?

- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [x] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added
- [ ] Documentation has been updated

tammy-baylis-swi commented on instrumentation/opentelemetry-instrumentation-asyncpg/src/opentelemetry/instrumentation/asyncpg/__init__.py:
```diff
@@ -150,8 +150,8 @@ def _uninstrument(self, **__):
-        params = getattr(instance, "_params", {})
-        name = args[0] if args[0] else params.get("database", "postgresql")
+        params = getattr(instance, "_params")
+        name = args[0] if args[0] else getattr(params, "database", "postgresql")
```
Nice. Could this also be updated in [_do_cursor_execute](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/317f7a6150d73f89959e8ec44c6d9f599977d289/instrumentation/opentelemetry-instrumentation-asyncpg/src/opentelemetry/instrumentation/asyncpg/__init__.py#L189-L193)?

privatwolke replied:
Done!

tammy-baylis-swi commented on instrumentation/opentelemetry-instrumentation-asyncpg/src/opentelemetry/instrumentation/asyncpg/__init__.py:
```diff
@@ -150,8 +150,8 @@ def _uninstrument(self, **__):
 
     async def _do_execute(self, func, instance, args, kwargs):
         exception = None
-        params = getattr(instance, "_params", {})
```
Would we still want to keep the default to prevent other AttributeError?

privatwolke replied:
Yup, fixed!

tammy-baylis-swi replied:
Thanks! We should still use the original default of `{}` instead of `None`, else the `getattr` call will give AttributeError for NoneType object if there's no `_params`.

privatwolke replied:
Not really - `getattr`works just fine on `None`:
```py
>>> getattr(None, '_params', 'bar')
'bar'
```

That means that if there is no `_params`, we would have `params = None`. After that we do `getattr(params, "database", "postgresql")` which is evaluated to `getattr(None, "database", "postgresql")` which finally evaluates to `"postgresql"`.

tammy-baylis-swi replied:
Yes, you are correct 👍

tammy-baylis-swi commented:
Thanks @privatwolke for revisiting this especially with the added test coverage.

To get [this check](https://github.com/open-telemetry/opentelemetry-python-contrib/actions/runs/13242594700/job/36974933833?pr=3253) to pass, please could you run `tox -e ruff` on your local then commit the changes.

privatwolke commented:
@tammy-baylis-swi - thanks for the quick review! I've updated the PR with the suggested changes and also added additional tests for the `Cursor` instrumentation. Code formatting should now pass as well ✨

privatwolke commented on instrumentation/opentelemetry-instrumentation-asyncpg/src/opentelemetry/instrumentation/asyncpg/__init__.py:
```diff
@@ -150,8 +150,10 @@ def _uninstrument(self, **__):
-        name = args[0] if args[0] else params.get("database", "postgresql")
+        params = getattr(instance, "_params", None)
+        name = (
+            args[0] if args[0] else getattr(params, "database", "postgresql")
```
TBH, I don't think that this code is possible to hit without doing some pretty shady things to the internals of asyncpg. It would mean that asyncpg has changed in a way that either `Connection._params` is not there anymore or that the `ConnectionParams` namedtuple has no `database` member anymore.

Even if you leave the `database` kwarg blank when you initiate the connection, it'll fall back on the username as the default database. In that case `database` is still a member of the `ConnectionParams` ([and has a value](https://github.com/MagicStack/asyncpg/blob/master/asyncpg/connect_utils.py#L464-L465)).

I've added a test for this using `mock.object` to change the `_params` property since I don't see any real-world scenario of this happening without upstream changes.

tammy-baylis-swi replied:
Yes, mocks work just fine for this!

privatwolke commented:
@xrmx I think something went wrong with the automerge. Should I rebase my branch on the base branch?
